src/access/AccessRestrictionProvider.h

```diff
@@ -0,0 +1,225 @@
+     *
+     * @param [in] listener  The listener to remove.
+     */
+    void RemoveListener(Listener & listener);
+
+    /**
+     * Check whether access by a subject descriptor to a request path should be restricted (denied) for the given action.
+     * These restrictions are are only a part of overall access evaluation.
+     *
+     * If access is not restricted, CHIP_NO_ERROR will be returned.
+     *
+     * @retval CHIP_ERROR_ACCESS_DENIED if access is denied.
+     * @retval other errors should also be treated as restricted/denied.
+     * @retval CHIP_NO_ERROR if access is not restricted/denied.
+     */
+    CHIP_ERROR Check(const SubjectDescriptor & subjectDescriptor, const RequestPath & requestPath);
+
+    /**
+     * Request a review of the access restrictions for a fabric.
+     *
+     * @param [in]  fabricIndex  The index of the fabric requesting a review.
+     * @param [in]  arl          An optinal list of access restriction entries to review.  If null, all entries will be reviewed.
+     * @param [out] token        The unique token for the review, which can be matched to a review update event.
+     */
+    CHIP_ERROR RequestFabricRestrictionReview(FabricIndex fabricIndex, const std::vector<Entry> & arl, uint64_t & token)
+    {
+        token = ++mNextToken;
+        return DoRequestFabricRestrictionReview(fabricIndex, token, arl);
+    }
+
+    /**
+     * Get the commissioning restriction entries.
+     *
+     * @retval the commissioning restriction entries.
+     */
+    const std::vector<Entry> & GetCommissioningEntries() const { return mCommissioningEntries; }
+
+    /**
+     * Get the restriction entries for a fabric.
+     *
+     * @param [in]  fabricIndex the index of the fabric for which to get entries.
+     * @param [out] entries reference to a const vector to hold the entries.
+     */
+    CHIP_ERROR GetEntries(const FabricIndex fabricIndex, const std::vector<Entry> *& entries) const
+    {
+        auto it = mFabricEntries.find(fabricIndex);
+        if (it == mFabricEntries.end())
+        {
+            return CHIP_ERROR_NOT_FOUND;
+        }
+
+        entries = &(it->second);
+
+        return CHIP_NO_ERROR;
+    }
+
+protected:
+    /**
+     * Initiate a review of the access restrictions for a fabric. This method should be implemented by the platform and be
+     * non-blocking.
+     *
+     * @param [in] fabricIndex  The index of the fabric requesting a review.
+     * @param [in] token        The unique token for the review, which can be matched to a review update event.
+     * @param [in] arl          An optinal list of access restriction entries to review.  If null, all entries will be reviewed.
+     * @return                  CHIP_NO_ERROR if the review was successfully requested, or an error code if the request failed.
+     */
+    virtual CHIP_ERROR DoRequestFabricRestrictionReview(const FabricIndex fabricIndex, uint64_t token,
```

Comment:
So the way this is currently implemented, the client can get the event before it gets the command response.  This is fine, and unavoidable, but the spec should probably have some informative language making this clear to spec readers.

Reply:
Resolving and tracking with https://github.com/project-chip/connectedhomeip/issues/35165